### PR TITLE
Add `Base.checked_pow(x,y)` to `Base.Checked` library

### LINKED ISSUE #52262: No `Base.Checked.checked_pow` function
There are a bunch of checked arithmetic functions that throw an OverflowError on overflow.

There isn't currently a `checked_pow` which implements `^` with the OverflowError check. Can we add one?

I think it would delegate to `Base.checked_mul` in its while-loop, rather than how `power_by_squaring` currently calls `*`. Something like:
```julia
function checked_power_by_squaring(x_, p::Integer)
...
    elseif p == 2
        return Base.checked_mul(x, x)
...
    while (t -= 1) > 0
        x = Base.checked_mul(x, x)
    end
...
end
```
etc

### PR DESCRIPTION
**Pull request by NHDaly**
Fixes #52262.

Performs `^(x, y)` but throws OverflowError on overflow.

Example:
```julia
julia> 2^62
4611686018427387904

julia> 2^63
-9223372036854775808

julia> checked_pow(2, 63)
ERROR: OverflowError: 2147483648 * 4294967296 overflowed for type Int64
```

### CODE REVIEW

File: base/intfuncs.jl [hunk 1]
```diff
@@ -309,14 +309,16 @@ to_power_type(x) = convert(Base._return_type(*, Tuple{typeof(x), typeof(x)}), x)
     "\nMake x a float matrix by adding a zero decimal ",
     "(e.g., [2.0 1.0;1.0 0.0]^", p, " instead of [2 1;1 0]^", p, ")",
     "or write float(x)^", p, " or Rational.(x)^", p, ".")))
-@assume_effects :terminates_locally function power_by_squaring(x_, p::Integer)
+power_by_squaring(x_, p::Integer) = power_by_squaring_with_op(*, x_, p)
```
Comment (StefanKarpinski) [inline base/intfuncs.jl:312]: How about just changing this to `power_by_squaring(x_, p::Integer; *=*)` and then the rest of the function stays the same?

File: base/checked.jl [hunk 2]
```diff
@@ -358,6 +358,21 @@ The overflow protection may impose a perceptible performance penalty.
...
+    Base.checked_pow(x, y)
+
+Calculates `^(x,y)`, checking for overflow errors where applicable.
+
+The overflow protection may impose a perceptible performance penalty.
+"""
+checked_pow(x::Integer, y::Integer) = checked_power_by_squaring(x, y)
+
+checked_power_by_squaring(x_, p::Integer) =
+    Base.power_by_squaring_with_op(checked_mul, x_, p)
```
Comment (StefanKarpinski) [inline base/checked.jl:371]: Then this just becomes:
```suggestion
    Base.power_by_squaring_with_op(x_, p, * = checked_mul)
```

File: base/intfuncs.jl (hunk 1 above)
Comment (NHDaly) [reply base/intfuncs.jl:312]: Clever! Sure that sounds fine to me. Are there any perf issues with kwargs these days for a tight mathematical op like `^`?

Comment (oscardssmith) [reply base/intfuncs.jl:312]: don't you have to do something to make it specialize on the `mul` operation?

Comment (NHDaly) [reply base/intfuncs.jl:312]: I think as long as you call the function, we do specialize on function args, oscar

File: base/checked.jl (hunk 2 above)
Comment (NHDaly) [reply base/checked.jl:371]: nice. done!

File: base/intfuncs.jl (hunk 1 above)
Comment (NHDaly) [reply base/intfuncs.jl]: k thanks. it's a bit mad, but i like it. 😁 i've applied this suggestion 👍

Comment (NHDaly) [reply base/intfuncs.jl]: Oh, sadly, `*` is not an allowed variable name:
```julia
ERROR: ParseError:
# Error @ REPL[1]:1:80
@assume_effects :terminates_locally function power_by_squaring(x_, p::Integer; *=*)
#                                                                              └┘ ── invalid identifier
```
So i've pushed up one more commit here using `mul`

Comment (mbauman) [reply base/intfuncs.jl]: With colors it becomes a bit more obvious that this is a whitespace thing:

image

Kinda funny that it's trying to parse `*=` as the identifier, but spacing things out (or parentheses or both) works.

Comment (StefanKarpinski) [reply base/intfuncs.jl]: It's totally allowed as a variable name:
```jl
julia> * = "Hello"
"Hello"
```
... and it works if you use spaces:
```jl
julia> f(x; * = *) = x*x
f (generic function with 1 method)

julia> f(3)
9
```

Comment (StefanKarpinski) [reply base/intfuncs.jl]: The parsing issue is that `*=` is an operator, which is why you need a space.

Comment (oscardssmith) [reply base/intfuncs.jl]: `(*) = (*)` is also an interesting face

**Review by oscardssmith — APPROVED** [review]

**Review by inkydragon — COMMENTED** [review]
Don't forget to add the new function to the doc.
https://github.com/JuliaLang/julia/blob/cf44b530a4499825ba4b7da5d2eccc1611e8c14b/doc/src/base/math.md?plain=1#L153-L154

File: base/intfuncs.jl (hunk 1 above)
Comment (NHDaly) [reply base/intfuncs.jl]: hah, i missed that -- i just assumed i was misreading the parser's highlighting. I haven't spent much time with the new parser yet and i'm not used to it - still working mostly in 1.9. Thanks all

**Comment by inkydragon** [comment]
Test `compiler/ssair` failed on [test aarch64-apple-darwin](https://buildkite.com/julialang/julia-master/builds/32921#018d5c82-8286-4fc6-a77f-101bd1ef252e)

Looks rarely seen, not sure if it's related.

<details>
<summary>test log</summary>

```
The global RNG seed was 0xaa6cce87493099399c5d4c23b781e1.
Error in testset compiler/ssair:
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XG3Q6T6R70.0/build/default-honeycrisp-XG3Q6T6R70-0/julialang/julia-master/julia-ad6f0489c3/share/julia/test/compiler/ssair.jl:552
  Expression: length(ir.stmts) == 2
   Evaluated: 3 == 2
Error in testset compiler/ssair:
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XG3Q6T6R70.0/build/default-honeycrisp-XG3Q6T6R70-0/julialang/julia-master/julia-ad6f0489c3/share/julia/test/compiler/ssair.jl:553
  Expression: Meta.isexpr((ir.stmts[1])[:stmt], :invoke)
Error in testset compiler/ssair:
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XG3Q6T6R70.0/build/default-honeycrisp-XG3Q6T6R70-0/julialang/julia-master/julia-ad6f0489c3/share/julia/test/compiler/ssair.jl:559
  Expression: length(ir.stmts) == 4
   Evaluated: 5 == 4
Error in testset compiler/ssair:
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XG3Q6T6R70.0/build/default-honeycrisp-XG3Q6T6R70-0/julialang/julia-master/julia-ad6f0489c3/share/julia/test/compiler/ssair.jl:560
  Expression: Meta.isexpr((ir.stmts[1])[:stmt], :invoke)
Error in testset compiler/ssair:
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XG3Q6T6R70.0/build/default-honeycrisp-XG3Q6T6R70-0/julialang/julia-master/julia-ad6f0489c3/share/julia/test/compiler/ssair.jl:632
  Expression: new_call_idx !== nothing
   Evaluated: nothing !== nothing
Error in testset compiler/ssair:
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XG3Q6T6R70.0/build/default-honeycrisp-XG3Q6T6R70-0/julialang/julia-master/julia-ad6f0489c3/share/julia/test/compiler/ssair.jl:633
  Expression: new_call_idx == new_invoke_idx + 1
   Evaluated: nothing == 4

```

</details>

**Comment by vtjnash** [comment]
That seems definitely related, since it is a test for inferability of `^` that this changes. Maybe @aviatesk can take a quick look and see if something is going wrong there, or if the test is wrong?

**Comment by aviatesk** [comment]
The modifications in this PR seem to be reasonable. I resolved test failures by improving the robustness of `compiler/ssair`.

**Comment by inkydragon** [comment]
REPL test failure not related, fixed by #53134. pr need rebase.

**Pull request merged**